Client/Frontend/Browser/Favicons/FaviconFetcher.swift

```diff
@@ -0,0 +1,494 @@
+                    func updateScore(x: Int, y: Int) {
+                        let location = ((Int(icon.size.width) * y) + x) * 4
+                        guard location + 3 < length else { return }
+                        let alpha = data[location + 3]
+                        if alpha == 255 {
+                            score -= 1
+                        } else {
+                            score += 1
+                        }
+                    }
+                    for x in 0..<Int(icon.size.width) {
+                        updateScore(x: x, y: 0)
+                    }
+                    for x in 0..<Int(icon.size.width) {
+                        updateScore(x: x, y: Int(icon.size.height))
+                    }
+                    // We've already scanned the first and last pixel during
+                    // top/bottom pass
+                    for y in 1..<Int(icon.size.height)-1 {
+                        updateScore(x: 0, y: y)
+                    }
+                    for y in 1..<Int(icon.size.height)-1 {
+                        updateScore(x: Int(icon.size.width), y: y)
+                    }
+                    DispatchQueue.main.async {
+                        completion(score > 0)
+                    }
+                }
+            } else {
+                DispatchQueue.main.async {
+                    completion(false)
+                }
+            }
+        }
+    }
+}
+
+private struct AssociatedObjectKeys {
+    static var faviconFetcher: Int = 0
+    static var monogramLabel: Int = 0
+}
+
+extension UIImageView {
+    
+    private var faviconFetcher: FaviconFetcher? {
+        get { objc_getAssociatedObject(self, &AssociatedObjectKeys.faviconFetcher) as? FaviconFetcher }
+        set { objc_setAssociatedObject(self, &AssociatedObjectKeys.faviconFetcher, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC) }
+    }
+    
+    private var monogramLabel: UILabel? {
+        get { objc_getAssociatedObject(self, &AssociatedObjectKeys.monogramLabel) as? UILabel }
+        set { objc_setAssociatedObject(self, &AssociatedObjectKeys.monogramLabel, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC) }
+    }
+    
+    /// Load the favicon from a site URL directly into a `UIImageView`. If no
+    /// favicon is found, a monogram will be used where the letter is determined
+    /// based on `fallbackMonogramCharacter` or the site URL.
+    ///
+    /// This method will only attempt to fetch a small favicon for a site.
+    /// Do not use this method if you need to ensure that a large apple-touch
+    /// icon is used.
+    func loadFavicon(for siteURL: URL,
+                     domain: Domain? = nil,
+                     fallbackMonogramCharacter: String? = nil,
+                     completion: (() -> Void)? = nil) {
+        monogramLabel?.removeFromSuperview()
+        faviconFetcher = FaviconFetcher(siteURL: siteURL, kind: .favicon, domain: domain)
+        faviconFetcher?.load { [weak self] url, attributes in
```

Comment:
I had it here basically to guard against reuse but you're right in this case we don't reuse `FaviconFetcher` since its re-initialized each time so we can just remove the url check